### PR TITLE
Time budget should not depend on the real clock

### LINKED ISSUE #95875: timeBudget shouldn't depend on the real clock
Instead of relying on the clock:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go#L60

 we should do something similar to the golang rate-limitter is doing:
https://github.com/kubernetes/kubernetes/blob/master/vendor/golang.org/x/time/rate/rate.go#L306-L331

### PR DESCRIPTION
**Pull request by aojea**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Insteado of relying on the clock and having a goroutine refreshing the budget,  estimate the budget based on the last access.

#### Which issue(s) this PR fixes:

Fixes #95875

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go [hunk 1]
```diff
@@ -19,11 +19,14 @@ package cacher
...
 	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
 )
 
 const (
-	refreshPerSecond = 50 * time.Millisecond
-	maxBudget        = 100 * time.Millisecond
+	refreshPerInterval = 50 * time.Millisecond
+	refreshInterval    = 1 * time.Second
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go:28]: I suggest simplifying a bit and just changing to just having "refreshPerSecond".

refreshInternal is just a multiplication factor for that, so you can achieve the same by having just refreshPerSecond.

Same below in timeBudgetImpl.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go [hunk 2]
```diff
@@ -46,42 +49,43 @@ type timeBudget interface {
...
-}
-
 func (t *timeBudgetImpl) takeAvailable() time.Duration {
 	t.Lock()
 	defer t.Unlock()
+	// budget accumulated since last access
+	acc := t.clock.Since(t.last).Seconds() * (t.refresh.Seconds() / t.refreshInterval.Seconds())
+	// this shouldn't be possible
+	if acc < 0 {
+		return time.Duration(0)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go:82]: This isn't correct - we should return t.budget, right?

What you need here is to simply set acc = 0 in this case and proceed.

[BTW - a unit test for this case would be useful]

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go [hunk 3]
```diff
@@ -46,42 +49,43 @@ type timeBudget interface {
...
+	}
+	// update current budget and store the current time
+	if t.budget = t.budget + time.Duration(acc*1e9); t.budget > t.maxBudget {
+		t.budget = t.maxBudget
+	}
+	t.last = t.clock.Now()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go:88]: nit: I think that we should be using the same now in all computations. So basically:
```
now := t.clock.Now()
acc := now.Sub(t.last).Seconds * ...
...
t.last = now
```

**Review by wojtek-t — COMMENTED** [review]
Thanks for picking this up!

**Comment by aojea** [comment]
/test pull-kubernetes-integration
I think that the failures are unrelated,

**Comment by wojtek-t** [comment]
This is great - thanks a lot!

/lgtm
/approve

**Pull request merged**